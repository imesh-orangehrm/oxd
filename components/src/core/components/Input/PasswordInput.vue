<template>
  <div :class="strengthClass">
    <oxd-text class="password-strength-check" v-if="showStrength">
      {{ strengthLabel }}
    </oxd-text>
    <oxd-input
      v-bind="$attrs"
      class="password-input"
      :hasError="hasError"
      :disabled="disabled"
      :readonly="readonly"
      :type="isPasswordVisible ? 'text' : 'password'"
    ></oxd-input>
    <div
      :class="[
        'password-view-icon-container',
        {'password-visible': isPasswordVisible},
      ]"
      @click="togglePasswordVisibility"
    >
      <oxd-icon-button
        class="oxd-password-view-icon"
        :name="isPasswordVisible ? 'eye-slash-fill' : 'eye-fill'"
        :iconStyles="isPasswordVisible ? {color: 'white'} : {}"
        :tooltip="isPasswordVisible ? 'Hide Password' : 'Show Password'"
        size="large"
      />
    </div>
  </div>
</template>
<script lang="ts">
import {defineComponent, ref, computed} from 'vue';
import Input from '@orangehrm/oxd/core/components/Input/Input.vue';
import oxdText from '@orangehrm/oxd/core/components/Text/Text.vue';
import IconButton from '@orangehrm/oxd/core/components/Button/Icon.vue';
import useTranslate from '@orangehrm/oxd/composables/useTranslate';

export default defineComponent({
  name: 'oxd-password-input',

  components: {
    'oxd-input': Input,
    'oxd-text': oxdText,
    'oxd-icon-button': IconButton,
  },

  props: {
    strength: {
      type: Number,
      default: -1,
      validator: (value: number) => [-1, 0, 1, 2, 3, 4, 5].includes(value),
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    readonly: {
      type: Boolean,
      default: false,
    },
    hasMinimumPasswordStrength: {
      type: Boolean,
      default: true,
    },
  },

  setup(props) {
    // Reactive reference for password visibility
    const isPasswordVisible = ref(false);
    const {$t} = useTranslate();

    // Computed properties for strengthClass, strengthLabel, and showStrength
    const strengthClass = computed(() => {
      const strengthClasses = [
        'very-weak-strength',
        'weak-strength',
        'better-strength',
        'medium-strength',
        'strong-strength',
        'strongest-strength',
      ];

      if (!props.hasMinimumPasswordStrength) {
        return 'password-container below-minimum-strength';
      }

      if (
        props.strength === -1 ||
        props.hasError ||
        props.disabled ||
        props.readonly
      ) {
        return 'password-container';
      }

      return `password-container ${strengthClasses[props.strength]}`;
    });

    const strengthLabel = computed(() => {
      const strengthLabels = [
        'Very Weak',
        'Weak',
        'Better',
        'Medium',
        'Strong',
        'Strongest',
      ];
      return $t(strengthLabels[props.strength]);
    });

    const showStrength = computed(() => {
      if (!props.hasMinimumPasswordStrength && props.strength !== -1) {
        return true;
      }
      return (
        props.strength !== -1 &&
        !props.hasError &&
        !props.disabled &&
        !props.readonly
      );
    });

    // Method to toggle password visibility
    const togglePasswordVisibility = () => {
      isPasswordVisible.value = !isPasswordVisible.value;
    };

    return {
      isPasswordVisible,
      strengthClass,
      strengthLabel,
      showStrength,
      togglePasswordVisibility,
    };
  },
});
</script>
<style src="./password-input.scss" lang="scss" scoped></style>
