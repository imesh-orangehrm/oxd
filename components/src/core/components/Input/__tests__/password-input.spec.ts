import {mount} from '@vue/test-utils';
import PasswordInput from '@orangehrm/oxd/core/components/Input/PasswordInput.vue';

describe('PasswordInput.vue', () => {
  it('renders OXD Input', () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Input with custom color', () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, style: {backgroundColor: 'aliceblue'}},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Input with error', () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, hasError: true},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Input with error message', () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, hasError: true, errorMessage: 'Error message'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Input with custom placeholder', () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, placeholder: 'Custom placeholder'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Input with strength meter', () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, strength: 1},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('if stregth value is 1 then it should show medium', async () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, strength: 1},
    });
    await wrapper.find('.password-input').setValue('oldPass123');
    expect(wrapper.html()).toContain('Weak');
  });

  it('if stregth value is 2 then it should show medium', async () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, strength: 2},
    });
    await wrapper.find('.password-input').setValue('oldPass123');
    expect(wrapper.html()).toContain('Better');
  });
  it('if stregth value is 3 then it should show strong', async () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, strength: 3},
    });
    await wrapper.find('.password-input').setValue('oldPass123');
    expect(wrapper.html()).toContain('Medium');
  });
  it('if stregth value is 4 then it should show very strong', async () => {
    const value = 'Input';
    const wrapper = mount(PasswordInput, {
      props: {label: value, strength: 4},
    });
    await wrapper.find('.password-input').setValue('oldPass123');
    expect(wrapper.html()).toContain('Strong');
  });
  it('on input', async () => {
    const wrapper = mount(PasswordInput, {});
    const input = wrapper.find('input');
    await input.trigger('input');
    expect(wrapper.emitted()).toHaveProperty('input');
  });
  it('should toggle password visibility when icon is clicked', async () => {
    const wrapper = mount(PasswordInput);
    expect(wrapper.vm.isPasswordVisible).toBe(false);
    const toggleButton = wrapper.find('.password-view-icon-container');
    await toggleButton.trigger('click');
    expect(wrapper.vm.isPasswordVisible).toBe(true);
    await toggleButton.trigger('click');
    expect(wrapper.vm.isPasswordVisible).toBe(false);
  });
  it('should hide password strength when password is too short', async () => {
    const wrapper = mount(PasswordInput, {
      data() {
        return {
          password: 'short',
        };
      },
    });

    expect(wrapper.vm.showStrength).toBe(false);
  });
  it('should show password strength when password is long enough', async () => {
    const wrapper = mount(PasswordInput, {
      props: {
        strength: 1,
        hasError: false,
      },
      data() {
        return {
          password: 'longEnoughPassword',
        };
      },
    });

    expect(wrapper.vm.showStrength).toBe(true);
  });
  it('should hid password strength when password is too long ', async () => {
    const wrapper = mount(PasswordInput, {
      data() {
        return {
          password: 'a'.repeat(65),
        };
      },
    });

    expect(wrapper.vm.showStrength).toBe(false);
  });
  it('should hide password strength when component is disabled', async () => {
    const wrapper = mount(PasswordInput, {
      props: {
        disabled: true,
      },
    });

    expect(wrapper.vm.showStrength).toBe(false);
  });
  it('should hide password strength when component is readonly', async () => {
    const wrapper = mount(PasswordInput, {
      props: {
        readonly: true,
      },
    });

    expect(wrapper.vm.showStrength).toBe(false);
  });
  it('should hide password strength when component has an error', async () => {
    const wrapper = mount(PasswordInput, {
      props: {
        hasError: true,
      },
    });

    expect(wrapper.vm.showStrength).toBe(false);
  });
  it('should hide password strength when component has no strength prop', async () => {
    const wrapper = mount(PasswordInput, {});
    expect(wrapper.vm.showStrength).toBe(false);
  });
  it('should show password strength gray color when hasMinimumPasswordStrength is false', async () => {
    const wrapper = mount(PasswordInput, {
      props: {
        hasMinimumPasswordStrength: false,
        strength: 1,
      },
    });
    expect(wrapper.vm.showStrength).toBe(true);
    expect(wrapper.vm.strengthClass).toBe(
      'password-container below-minimum-strength',
    );
  });
});
