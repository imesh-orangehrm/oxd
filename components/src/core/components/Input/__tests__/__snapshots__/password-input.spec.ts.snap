// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput.vue renders OXD Input 1`] = `
<div class="password-container" label="Input">
  <!--v-if-->
  <div class="input-outer-wrapper">
    <!--v-if-->
    <!--v-if--><input class="oxd-input oxd-input--active password-input" label="Input" type="password">
  </div>
  <div class="password-view-icon-container"><button type="button" class="oxd-icon-button oxd-password-view-icon" tooltip="Show Password" flow="top"><i class="oxd-icon oxd-icon--large bi-eye-fill"></i></button></div>
</div>
`;

exports[`PasswordInput.vue should renders OXD Input with custom color 1`] = `
<div class="password-container" label="Input" style="background-color: aliceblue;">
  <!--v-if-->
  <div class="input-outer-wrapper">
    <!--v-if-->
    <!--v-if--><input class="oxd-input oxd-input--active password-input" style="background-color: aliceblue;" label="Input" type="password">
  </div>
  <div class="password-view-icon-container"><button type="button" class="oxd-icon-button oxd-password-view-icon" tooltip="Show Password" flow="top"><i class="oxd-icon oxd-icon--large bi-eye-fill"></i></button></div>
</div>
`;

exports[`PasswordInput.vue should renders OXD Input with custom placeholder 1`] = `
<div class="password-container" label="Input" placeholder="Custom placeholder">
  <!--v-if-->
  <div class="input-outer-wrapper">
    <!--v-if-->
    <!--v-if--><input class="oxd-input oxd-input--active password-input" label="Input" placeholder="Custom placeholder" type="password">
  </div>
  <div class="password-view-icon-container"><button type="button" class="oxd-icon-button oxd-password-view-icon" tooltip="Show Password" flow="top"><i class="oxd-icon oxd-icon--large bi-eye-fill"></i></button></div>
</div>
`;

exports[`PasswordInput.vue should renders OXD Input with error 1`] = `
<div class="password-container" label="Input">
  <!--v-if-->
  <div class="input-outer-wrapper">
    <!--v-if-->
    <!--v-if--><input class="oxd-input oxd-input--active oxd-input--error password-input" label="Input" type="password">
  </div>
  <div class="password-view-icon-container"><button type="button" class="oxd-icon-button oxd-password-view-icon" tooltip="Show Password" flow="top"><i class="oxd-icon oxd-icon--large bi-eye-fill"></i></button></div>
</div>
`;

exports[`PasswordInput.vue should renders OXD Input with error message 1`] = `
<div class="password-container" label="Input" errormessage="Error message">
  <!--v-if-->
  <div class="input-outer-wrapper">
    <!--v-if-->
    <!--v-if--><input class="oxd-input oxd-input--active oxd-input--error password-input" label="Input" errormessage="Error message" type="password">
  </div>
  <div class="password-view-icon-container"><button type="button" class="oxd-icon-button oxd-password-view-icon" tooltip="Show Password" flow="top"><i class="oxd-icon oxd-icon--large bi-eye-fill"></i></button></div>
</div>
`;

exports[`PasswordInput.vue should renders OXD Input with strength meter 1`] = `
<div class="password-container weak-strength" label="Input">
  <p class="oxd-text oxd-text--p password-strength-check">Weak</p>
  <div class="input-outer-wrapper">
    <!--v-if-->
    <!--v-if--><input class="oxd-input oxd-input--active password-input" label="Input" type="password">
  </div>
  <div class="password-view-icon-container"><button type="button" class="oxd-icon-button oxd-password-view-icon" tooltip="Show Password" flow="top"><i class="oxd-icon oxd-icon--large bi-eye-fill"></i></button></div>
</div>
`;
